### ISSUE TITLE
"And I pause scenario execution" doesn't work

### ISSUE THREAD

**Issue by tmuras** [moodlehq/moodle-docker#96]
Using the step "And I pause scenario execution" triggers an exception:

`Break point should only be used with interative terminal. (Behat\Mink\Exception\ExpectationException)`

even it the tests are run from inside docker container, using interactive session.

The good news is that it's only the check that is not working in docker:

```
        // Make sure this step is only used with interactive terminal (if detected).
        if ($posixexists && !@posix_isatty(STDOUT)) {
            $session = $this->getSession();
             throw new ExpectationException('Break point should only be used with interative terminal.', $session);
        }

```

Simply commenting out throwing an exception line will make the step to behave correctly:
![image](https://user-images.githubusercontent.com/695281/51526702-3998ee80-1e33-11e9-8d36-8bb66d3a0ce9.png)

**Comment by scara** [comment]
Hi @tmuras,
never tried by myself but it looks like we are missing `-i` or `-t` or `-it` here i.e. something like:
```yml
stdin_open: true # Maps -i
#tty: true # Maps -t, In case posix_isatty() requires even a TTY since it is called against STDOUT
```
for the composition i.e.:
```diff
$ git diff
diff --git a/base.yml b/base.yml
index 24f3882..6b7cdc0 100755
--- a/base.yml
+++ b/base.yml
@@ -2,6 +2,8 @@ version: "2"
 services:
   webserver:
     image: "moodlehq/moodle-php-apache:${MOODLE_DOCKER_PHP_VERSION}"
+    stdin_open: true
+    #tty: true
     depends_on:
       - db
     volumes:
```

It would be a nice add-on if it will work, given the right combination (just `-i` or `-it` or `-t`).

HTH,
Matteo

**Comment by paulholden** [comment]
Thanks @stronk7 for referring me to this task. I can confirm that @tmuras workaround to comment out the thrown exception during an interaction session on the Docker container did indeed work (after connecting via `docker exec -it docker_webserver_1 bash`)

**Comment by sammarshallou** [comment]
I'm hitting this problem in moodle-docker setup, using a command like:

`moodle-docker-compose exec webserver php admin/tool/behat/cli/run.php --tags=$1`

The interesting point is that if I add this code near the start of behat/cli/run.php:

`echo(posix_isatty(STDOUT) ? "YES\n" : "No\n");`

It echoes YES. However, the same code in the part where it tests for pause outputs 'No'. I assume the difference is because Behat messes about with STDOUT, but I don't know exactly why.

**Comment by tmuras** [comment]
I have run it by claude and got this:
  At the start of run.php: The script runs directly in your terminal, so STDOUT is connected to your TTY. posix_isatty(STDOUT) returns true → "YES".                                          
                                                                                                                                                                                              
  When the pause function runs: The behat tests are executed in child processes spawned via Symfony's Process component (see lib/behat/lib.php:547):                                          
                                                                                                                                                                                              
  $process = Symfony\Component\Process\Process::fromShellCommandline($cmd);                                                                                                                   
  ...                                                                                                                                                                                         
  $processes[$name]->start();                                                                                                                                                                 
                                                                                                                                                                                              
  Symfony Process uses pipes (not a PTY) to connect to the child process's STDOUT so the parent can capture the output via $process->getOutput() and $process->getIncrementalOutput(). The    
  child's STDOUT is connected to a pipe, not the terminal.                                                                                                                                    
                                                                                                                                                                                              
  So the flow is:                                                                                                                                                                             
  - Your terminal → run.php (STDOUT=TTY) → spawns child → behat process (STDOUT=pipe) → parent captures output                                                                                
                                                                                                                                                                                              
  When the pause function in lib/behat/classes/util.php:489 checks posix_isatty(STDOUT), it's running inside the child process where STDOUT is a pipe, not the terminal - hence "No".         
                                                                                                                                                                                              
  This is actually intentional. The pause check exists to prevent breakpoints from being used in non-interactive contexts (like CI pipelines or parallel runs) where there's no human to press
   a key. If you want to use breakpoints, you'd need to run behat in single-run mode without parallelization, directly invoking the behat binary rather than going through run.php's parallel 
  execution.

**Comment by NeillM** [comment]
I also tried running behat via:

`moodle-docker-compose exec webserver vendor/bin/behat --config /var/www/behatdata/behatrun/behat/behat.yml  --tags=$1`

and that had the same issue.

I was running moodle-docker via an Ubuntu image in WSL.

**Comment by sammarshallou** [comment]
> This is actually intentional. The pause check exists to prevent breakpoints from being used in non-interactive contexts (like CI pipelines or parallel runs) where there's no human to press a key. If you want to use breakpoints, you'd need to run behat in single-run mode without parallelization, directly invoking the behat binary rather than going through run.php's parallel execution.

Thanks to Neil for confirming but just to clarify, I was using run.php in single (not parallel) mode. In other words it's doing the 'if (empty($parallelrun))' branch on line 219 of behat/cli/run.php, and does not run the cli_execute_parallel function which launches extra processes as mentioned.

And as mentioned above, if I comment out the 'throw' it does work correctly, i.e. I am using an interactive termainal, I see the pause message and when I press return it continues the test. 

So I don't think there is anything intentional about this behaviour - I agree it would work correctly if using parallel run.

I was fairly sure behat does something weird with stdout even when you're not doing parallel runs, but having searched the code a bit, I can't find it... I will try harder later today if I get time.

**Comment by NeillM** [comment]
I also use single runs on my dev box and get the pause failures using `behat/cli/run.php`

**Comment by sammarshallou** [comment]
I think this can be resolved in Moodle and is not a moodle-docker specific issue; I have filed https://moodle.atlassian.net/browse/MDL-87905 - We may be able to close it here if eveyone agrees.

**Comment by tmuras** [comment]
I agree Sam, seems like https://moodle.atlassian.net/browse/MDL-87905 is a good way to go.
